onset_maproom/calc.py

```diff
@@ -51,6 +51,29 @@ def water_balance(
     return water_balance
 
 
+def longest_spell(flagged_data, coord):
+    """Find the longest spell of flagged (0/1) data along coord
+    The longest spell is the maximum value
+    of the discrete difference
+    of cumulative flags
+    keeping only the unflagged data
+    Because diff needs at least 2 points,
+    we need to keep (where) the unflagged and first and last
+    with the cumulative value for last
+    and 0 for first
+    """
+    # Points to apply diff to
+    unflagged_and_ends = (flagged_data == 0) * 1
+    unflagged_and_ends[{coord: [0, -1]}] = 1
+    
+    ls = flagged_data.cumsum(coord).where(unflagged_and_ends, drop=True).where(
+        # first cumul point must be set to 0
+        lambda x: x[coord] != flagged_data[coord][0], other=0
```

Comment:
Possibly moot given my previous comment, but...

IIUC this is here because without it, if the 0th point is flagged then the 0th cumulative sum is 1, and consequently the length of the first run gets counted as n - 1 when it should be n.

This solution was hard for me to understand, and it's still not obvious to me that it's correct in all cases, even though I tried and failed to find a case where it's wrong. Wouldn't it be simpler to pad `coord` on the left with a slice of zeros, e.g. with `xr.concat`?

Reply:
You understand very well. It covers the case where the 1st day is flagged and contributes to the longest run. I am not able to explain it very well either. I think the best way to see it is that diff needs pairs so its result drops a data point and there is an origin issue.
I am not super excited about concatenating. Adding size to an object... plus is it not going to be a headache depending on the nature of the coord (regular step, or not...). Is it not going to create trouble when feeding this to a rolling?

Also, after I pushed, I realized that: I claimed that it would work for whatever dimension/step, but that is less clear to me now. The "length" that is computed is in deltas of steps. So if 9, it's 9 "steps." And that is likely true only for regular coords. If irregular, it's not 9 but it should be a sum of 9 deltas that vary... That brings back both the matters of working with units, and with coordinates having width, or not...

Maybe I should not try and make this that general....?

Reply:
> I think the best way to see it is that diff needs pairs so its result drops a data point and there is an origin issue.

Exactly, that's why adding an extra data point seems like the cleanest solution. I don't care enough to insist on it, though.

Reply:
you are right: I found cases that don't work. Concatenating with 0 is the right thing to do. However, after review, it turns out that doing nothing is equivalent to concatenating 0 (especially now that there is bfill).
If the first point is 0, it remains 0 through bfill, then either there were a run of n ones after and diff will return n-0 as we wish; and having concatenating 0 would have added the point 0-0 that doesn't interfere with max.
If the first point is 1, then whatever the rest of the series is, it's backfilled by bfill with the length of the run it belongs to, say n (n >= 1); then concatenating 0 preserves n = n-0 and we have what we want.

Thus morality: I just need to replace unflagged_and_ends with an unflagged_and_tail. That is still needed to cover the case where the last point is 1, so that the run it belongs to is taken into account. And at the same time it insures that there is at least 1 value, thus more than 1 after bfill, and thus diff can operate.

Reply:
I was wrong: i don't have a counter-example that doesn't work. And my proposed solution breaks the case where all are flagged, what shed some light on all this.

This:
"IIUC this is here because without it, if the 0th point is flagged then the 0th cumulative sum is 1, and consequently the length of the first run gets counted as n - 1 when it should be n."

is not correct. Or possibly is not correct anymore now that there is bfill. The consequence is that the length of the 1st spell would be counted as 0 (not n-1).

bfill propagates the value of the length of the first run, backwards, until there is a 0. If there is a 0, then that 0 remains 0 and diff gives the result we want. If there is no 0 (which also means that that first run starts on the first point), then the propagation goes all the way through and diff returns 0 for all those points. And thus, if it turned out that this was the longest run, max will miss it.

Consequently, concatenating an initial 0 covers that case, as our intuition tells us as we naturally understand pair-wise diff. Now, why is it equivalent to reset the first point to 0 before back filling? Because in that very case, we know the 1st point must be 1, thus the result of cumsum for the second point is at least 1 too (it's 1 or 2). Thus, the information in the 1st point must be contained in the 2nd and thus replacing the 1st point by 0 bears no consequences, it insures that the first run length is counted correctly, and it requires not use concat. That also means that the function is ambiguous in the case where the function is applied to a series of 1 point. If the series is 1 flagged point, then the where will try to replace that very point which is both ends at the same time by 0 and by 1. I don't know what it will do. Then diff will likely fail. Whereas the concat solution will return the right result in that very case.

Ok now. I am writing all this just to try to convince myself that I can explain what I code. I've been wrong before in my tentative explanations so I'm not 100% convinced by myself even now. Now I still have reservation with the concat solution (that we can however explain better), besides the fact that it increases the size of the object, is that won't I need to build a dim for that 0 with coords that somehow are anterior to the coord of the flagged_data? In other words I will have to make some assumption or get some info about the coord that so far I could ignore entirely? won't I?

Reply:
I pushed what I have so far. I guess either we are both convinced enough of what I wrote above, or not and I can be reasonable and implement concat solution

Reply:
[np.diff](https://numpy.org/doc/stable/reference/generated/numpy.diff.html) has an optional kwarg `prepend`, which looks like it exists for this exact reason.
```
>>> np.diff([1], prepend=0)
array([1])
```
But DataArray.diff doesn't support that option.

I still don't have an intuitive grasp of why your version works, but I worked through all the cases of length 2 and 3 and it gives the right answer every time, so I'm pretty convinced. You could add a special case for when the length of the sequence is 1.

Just for the sake of argument...

> won't I need to build a dim for that 0 with coords that somehow are anterior to the coord of the flagged_data? In other words I will have to make some assumption or get some info about the coord that so far I could ignore entirely? 

I think you could just drop the coord. Xarray tolerates dimensions without coordinates, and it doesn't seem like you need it (since you're only looking for the length of the longest run, not its location). Equivalently, you could extract the values as a numpy array and do the calculation in numpy (which has no notion of coordinates). But this is a moot point because your current implementation seems to work.